core/shared/src/test/scala/fs2/concurrent/ChannelSuite.scala

```diff
@@ -251,4 +251,88 @@ class ChannelSuite extends Fs2Suite {
 
     TestControl.executeEmbed(test)
   }
+
+  def checkIfSendAndCloseClosing(channel: IO[Channel[IO, Int]]) =
+    channel
+      .flatMap { ch =>
+        ch.sendAndClose(0) *> ch.isClosed
+      }
+      .assertEquals(true)
+
+  test("sendAndClose closes right after sending the last element in bounded(1) case") {
+    val channel = Channel.bounded[IO, Int](1)
+
+    checkIfSendAndCloseClosing(channel).parReplicateA_(if (isJVM) 10000 else 1)
+  }
+
+  test("sendAndClose closes right after sending the last element in bounded(2) case") {
+    val channel = Channel.bounded[IO, Int](2)
+
+    checkIfSendAndCloseClosing(channel).parReplicateA_(if (isJVM) 10000 else 1)
+  }
+
+  test("sendAndClose closes right after sending the last element in unbounded case") {
+    val channel = Channel.unbounded[IO, Int]
+
+    checkIfSendAndCloseClosing(channel).parReplicateA_(if (isJVM) 10000 else 1)
+  }
+
+  test("sendAndClose closes right after sending the last element in synchronous case") {
+    val test = Channel
+      .synchronous[IO, Int]
+      .flatMap { ch =>
+        ch.stream
+          .concurrently(Stream.emit(0).covary[IO].evalMap(ch.sendAndClose))
+          .compile
+          .drain *> ch.isClosed
+      }
+      .assertEquals(true)
+
+    test.parReplicateA_(if (isJVM) 10000 else 1)
+  }
+
+  def racingSendOperations(channel: IO[Channel[IO, Int]]) = {
+    val expectedFirstCase = ((Right(()), Right(())), List(0, 1))
+    val expectedSecondCase = ((Right(()), Left(Channel.Closed)), List(1))
+    val expectedThirdCase = ((Left(Channel.Closed), Right(())), List(1))
+
+    channel
+      .flatMap { ch =>
+        ch.send(0).both(ch.sendAndClose(1)).parProduct(ch.stream.compile.toList)
+      }
+      .map {
+        case obtained @ ((Right(()), Right(())), List(0, 1)) =>
+          assertEquals(obtained, expectedFirstCase)
+        case obtained @ ((Right(()), Left(Channel.Closed)), List(1)) =>
+          assertEquals(obtained, expectedSecondCase)
+        case obtained @ ((Left(Channel.Closed), Right(())), List(1)) =>
+          assertEquals(obtained, expectedThirdCase)
+        case e => fail(s"An unknown test result: $e")
+      }
+  }
+
+  test("racing send and sendAndClose should work in bounded(1) case") {
+    val channel = Channel.bounded[IO, Int](1)
+
+    racingSendOperations(channel).parReplicateA_(if (isJVM) 10000 else 1)
```

Comment:
Can we use `replicateA_` instead of `parReplicateA_`? I'm less worried about running the tests in parallel to each other and more interested about the fibers _within_ the test running parallel to each other. If we overwhelm the scheduler with 10000 of this test running at the same time then that is less likely to happen.

If it's slow we probably don't need this many replicates. There's only two possible outcomes: honestly maybe even a 100 replicates would suffice.

Oh, also we can probably move this to the `racingSendOperations` method so we don't have to repeat it for each test :)

Reply:
Yes, sure, a moment and I will refactor it